spark-operator-api/src/main/java/org/apache/spark/k8s/operator/status/AttemptInfo.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.k8s.operator.status;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class AttemptInfo {
+  @Builder.Default protected final Long id = 0L;
+
+  public AttemptInfo createNextAttemptInfo() {
```

Comment:
Is this thread-safe?

Reply:
No, this is not thread safe. It's to quickly create a new AttemptInfo based on existing one, e.g. when starting a new attempt.

Members of `ApplicationStatus` are not exposed via setter for the sake of integrity - and people can be mindful that a new `ApplicationStatus` need to be created in order to update status history or start a new attempt. For example

```
// ApplicationStatus need to include currentState, stateHistory, currentAttemptInfo, previousAttemptInfo
// when building status for a new attempt, currentAttempt becomes previous
AttemptInfo nextAttemptInfo = currentAttemptInfo.createNextAttemptInfo();
ApplicationStatus nextAttemptStatus = new ApplicationStatus(updatedState, updatedStateHistory, nextAttemptInfo, currentAttemptInfo)

```

Reply:
For all ID generation, we should be very carful. We cannot control the invoker-side.

We don't have any locking mechanism though out `api` module, do we?